src/gt_frontend/py_to_gtscript.py

```diff
@@ -0,0 +1,225 @@
+                        type_name, module.__name__
+                    )
+                )
+            return PyToGTScript._all_subclasses(getattr(module, type_name), module=module)
+        elif typ in [
+            eve.types.StrictStr,
+            eve.types.StrictInt,
+            eve.types.StrictFloat,
+            str,
+            int,
+            float,
+            type(None),
+        ]:  # todo: enhance
+            return set([typ])
+
+        raise ValueError("Invalid field type {}".format(typ))
+
+    # template is a 1-to-1 mapping from context and python ast node to gt4py ast node. context is encoded in the field types
+    # all understood sementic is encoded in the structure
+    class Templates:
+        Symbol = ast.Name(id=Capture("name"))
+
+        IterationOrder = ast.withitem(
+            context_expr=ast.Call(
+                func=ast.Name(id="computation"), args=[ast.Name(id=Capture("order"))]
+            )
+        )
+
+        Constant = ast.Constant(value=Capture("value"))
+
+        Interval = ast.withitem(
+            context_expr=ast.Call(
+                func=ast.Name(id="interval"), args=[Capture("start"), Capture("stop")]
+            )
+        )
+
+        LocationSpecification = ast.withitem(
+            context_expr=ast.Call(
+                func=ast.Name(id="location"), args=[ast.Name(id=Capture("location_type"))]
+            ),
+            optional_vars=Capture(
+                "name", default=ast.Name(id=UIDGenerator.get_unique_id(prefix="location"))
+            ),
+        )
+
+        SubscriptSingle = ast.Subscript(
+            value=Capture("value"), slice=ast.Index(value=ast.Name(id=Capture("index")))
+        )
+
+        SubscriptMultiple = ast.Subscript(
+            value=Capture("value"), slice=ast.Index(value=ast.Tuple(elts=Capture("indices")))
+        )
+
+        BinaryOp = ast.BinOp(op=Capture("op"), left=Capture("left"), right=Capture("right"))
+
+        Call = ast.Call(args=Capture("args"), func=ast.Name(id=Capture("func")))
+
+        LocationComprehension = ast.comprehension(
+            target=Capture("target"), iter=Capture("iterator")
+        )
+
+        Generator = ast.GeneratorExp(generators=Capture("generators"), elt=Capture("elt"))
+
+        Assign = ast.Assign(targets=[Capture("target")], value=Capture("value"))
+
+        Stencil = ast.With(items=Capture("iteration_spec"), body=Capture("body"))
+
+        Pass = ast.Pass()
+
+        Argument = ast.arg(arg=Capture("name"), annotation=Capture("type_"))
+
+        Computation = ast.FunctionDef(
+            args=ast.arguments(args=Capture("arguments")),
+            body=Capture("stencils"),
+            name=Capture("name"),
+        )
+
+    leaf_map = {
+        ast.Mult: gtc.common.BinaryOperator.MUL,
+        ast.Add: gtc.common.BinaryOperator.ADD,
+        ast.Div: gtc.common.BinaryOperator.DIV,
+    }
+
+    def transform(self, node, eligible_node_types=None):
```

Comment:
In the interest of readability and testability I would think about reducing the amount of nested code paths maybe by splitting some conditional paths into smaller functions.